NCCL process group: avoid workEnqueue when capturing cuda graph

bottler opened the pull request:
Summary:
In torch.distributed, we make ProcessGroupNCCL not call workEnqueue when the cuda stream is capturing. I.e., when capturing a CUDA graph, we do not enqueue anything for the watchdog thread to consider. This allows capturing NCCL operations in a CUDA Graph. 

This is followup to an internal discussion [1] where the watchdog thread was observed to crash when using cuda graphs containing an all_reduce. The watchdog thread wants to query events pertaining to enqueued work items, but this can't be done for "events" created during cuda graph capture.

[1] https://fb.workplace.com/groups/1405155842844877/posts/6975201909173548/

Test Plan: Test added. Also, the repro mentioned in https://fb.workplace.com/groups/1405155842844877/posts/7003002339726838/ runs successfully after this change.

Differential Revision: D46274814

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1605,7 +1605,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   work->store_ = store_;
 
-  if (!coalescing_state_) {
+  if (!coalescing_state_ && capture_status == c10::cuda::CaptureStatus::None) {
     workEnqueue(work);
```
I might prefer putting the capture status check inside `workEnqueue`, so that we could reject the enqueuing of any kind of collectives, P2P or coalesced collectives. Does that make sense?

bottler replied:
I think I haven't bypassed enough in all circumstances, hence the new test is giving an error in CI. I can try to fix that and to correct endCoalescing as well.

bottler replied:
Tests are now OK and I corrected endCoalescing. The check isn't inside workEnqueue though - I think it's simpler this way because in one case we already found the capture_status and in the other case we haven't, and also because it makes clear that more of the code could really be skipped when capturing.

kwen2501 reviewed (changes requested):
Thanks for the fix!
I just wonder if we could make a minor location change. Please see the inline comment. Thanks!

bottler committed:
nccl process group: no workEnqueue when capturing (#102542)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102542

When capturing a cuda graph, do not enqueue anything for the watchdog thread.

See discussion in https://fb.workplace.com/groups/1405155842844877/posts/6975201909173548/

Test Plan: The repro mentioned in https://fb.workplace.com/groups/1405155842844877/posts/7003002339726838/ runs successfully after this change.

Differential Revision: D46274814

fbshipit-source-id: 62a9d34343e2536cc566d1c636ba88a6c7c45b1b

kwen2501 reviewed (approved):
LGTM. Thanks for fixing the bug!

kwen2501 commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -m "Trunk failing tests consistently with this PR. Probably a land rance" -c landrace

Failing test:
https://hud.pytorch.org/failure/distributed%2Ftest_c10d_nccl.py%3A%3AProcessGroupNCCLTest%3A%3Atest_allreduce_in_cudagraph
